sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala

```diff
@@ -76,8 +101,12 @@ override def outputPartitioning: Partitioning = joinType match {
       rightIter: Iterable[InternalRow]): Iterator[InternalRow] = {
     val ret: Iterable[InternalRow] = {
       if (!key.anyNull) {
-        val temp = rightIter.collect {
-          case r if boundCondition(joinedRow.withRight(r)) => joinedRow.copy()
+        val temp = if (rightIter != null) {
+          rightIter.collect {
+            case r if boundCondition(joinedRow.withRight(r)) => joinedRow.copy()
+          }
+        } else {
+          List()
```

Comment:
Given that the old code seemed to make a special point of using an `EMPTY_LIST` constant I'm thinking that it may be  important to make sure we're not creating new objects here.